storage/metrics_storage.go

```diff
@@ -160,14 +171,116 @@ func (ms *MetricsStorageManager) consumeMetrics() {
+
+		messaged := false
+		var lastError error
+		for {
+			if (time.Now().After(timeout)) {
+				log.Errorf("Could not fetch the required secrets after 5 minutes. This may mean the secrets have not yet been created yet or the agent does not have the required permission to access the secret.");
+				os.Exit(1)
+			}
+
+			secret, err := client.Secrets(namespace).Get(secretName)
+			if err != nil {
+				if (lastError != err) {
+					log.Errorf("Error trying to get Secret named %v from namespace %v: %v", secretName, namespace, err)
+					lastError = err
+				}
+			} else {
+				value, found := secret.Data[key]
+				if !found {
+					if (lastError != err) {
+						log.Errorf("Secret %v does not contain key %v in namespace %v", secretName, key, namespace)
+						lastError = err
+					}
+				} else {
+					return value;
+				}
+			}
+
+			if (!messaged) {
+				log.Errorf("Could not get the requested secret. This may mean the secret has not yet been created yet or the agent does not have the required permission to access the secret. Will attempt again every 5 seconds for the next 5 minutes")
+			}
+			messaged = true
+			time.Sleep(5 * time.Second)
+		}
+	}
+
+	// function that will extract a credential string based on its value.
+	// If the string is prefixed with "secret:" it is assumed to be a key/value from a k8s secret.
+	// If the string is not prefixed, it is used as-is.
+	f := func(v string, file bool) string {
+		if strings.HasPrefix(v, "secret:") {
+			v = strings.TrimLeft(v, "secret:")
+			splits := strings.SplitN(v, "/", -1)
+			if len(splits) != 2 && len(splits) != 3 {
+				log.Errorf("Secret format is in valid for Hawkular Metrics: [%v]", v)
+				return ""
+			}
+
+			var namespace string
+			var secretName string
+			var secretKey string
+			if len(splits) == 2 {
+				namespace = conf.Kubernetes.Pod_Namespace
+				secretName = splits[0]
+				secretKey = splits[1]
+			} else {
+				namespace = splits[0]
+				secretName = splits[1]
+				secretKey = splits[2]
+			}
+			bytes := waitForSecret(namespace, secretName, secretKey)
+
+			if (!file) {
+				return strings.TrimSpace(string(bytes))
+			} else {
+				dirName := os.TempDir() + string(os.PathSeparator) + secretName
+				os.MkdirAll(dirName, 0744)
+				fileName := dirName + string(os.PathSeparator) + secretKey
+				fileErr := ioutil.WriteFile(fileName, bytes, 0644)
+				if fileErr != nil {
+					log.Errorf("Could not write the hawkular metric ca to file: %v", fileErr)
+				}
+				return fileName
+			}
+		} else {
+			return v
+		}
+	}
+
+
+	hawkularCredentials := config.Hawkular_Server{
+		URL: f(conf.Hawkular_Server.URL, false),
+		CA_Cert_File: f(conf.Hawkular_Server.CA_Cert_File, true),
+		Credentials: security.Credentials {
+			Password: f(conf.Hawkular_Server.Credentials.Password, false),
+			Username: f(conf.Hawkular_Server.Credentials.Username, false),
+			Token: f(conf.Hawkular_Server.Credentials.Token, false),
+		},
+		Tenant: f(conf.Hawkular_Server.Tenant, false),
```

Comment:
I never did address this in https://github.com/mwringe/hawkular-openshift-agent/pull/2 - I wasn't sure if this was needed or not, so I left it in assuming it is needed.